store/src/java/com/zimbra/cs/service/admin/HabOrgUnit.java

```diff
@@ -0,0 +1,179 @@
+import com.zimbra.cs.account.Domain;
+import com.zimbra.cs.account.Provisioning;
+import com.zimbra.cs.account.accesscontrol.AdminRight;
+import com.zimbra.cs.account.ldap.LdapObjectClass;
+import com.zimbra.cs.account.ldap.entry.LdapEntry;
+import com.zimbra.cs.ldap.LdapClient;
+import com.zimbra.cs.ldap.LdapException;
+import com.zimbra.cs.ldap.LdapServerType;
+import com.zimbra.cs.ldap.LdapUsage;
+import com.zimbra.cs.ldap.ZAttributes;
+import com.zimbra.cs.ldap.ZLdapContext;
+import com.zimbra.cs.ldap.ZLdapFilter;
+import com.zimbra.cs.ldap.ZLdapFilterFactory;
+import com.zimbra.cs.ldap.ZLdapFilterFactory.FilterId;
+import com.zimbra.cs.ldap.ZSearchControls;
+import com.zimbra.cs.ldap.ZSearchResultEnumeration;
+import com.zimbra.cs.ldap.ZSearchScope;
+import com.zimbra.cs.service.account.ToXML;
+import com.zimbra.soap.ZimbraSoapContext;
+
+
+/**
+ * @author zimbra
+ *
+ */
+public class HabOrgUnit extends AdminDocumentHandler {
+    
+
+    /* (non-Javadoc)
+     * @<HabOrgUnit op=“create/delete/rename” name=“”>
+     * <domain by=“name/id”></domain>
+     * </HabOrgUnit>
+     */
+    @Override
+    public Element handle(Element request, Map<String, Object> context) throws ServiceException {
+        ZimbraSoapContext zsc = getZimbraSoapContext(context);
+        Provisioning prov = Provisioning.getInstance();
+
+        Element d = request.getElement(AdminConstants.E_DOMAIN);
+        String domainKey = d.getAttribute(AdminConstants.A_BY);
+        String domainValue = d.getText();
+        
+        Domain domain = prov.get(Key.DomainBy.fromString(domainKey), domainValue);
+        if (domain == null) {
+            throw AccountServiceException.NO_SUCH_DOMAIN(domainValue);
+        }
+        checkDomainRight(zsc, domain, AdminRight.PR_ALWAYS_ALLOW);
+        String domainDn = ((LdapEntry)domain).getDN();
+        
+        
+        String habOrgUnitName = request.getAttribute(AdminConstants.A_NAME);
+        if (StringUtil.isNullOrEmpty(habOrgUnitName) ) {
+            throw ServiceException.INVALID_REQUEST(
+                String.format("Hab Org unit name is required, requested name:%s",habOrgUnitName), null);
+        }
+        
+        String operation = request.getAttribute(AdminConstants.A_OPERATION);
+        if (StringUtil.isNullOrEmpty(operation) ) {
+            throw ServiceException.INVALID_REQUEST(
+                String.format("Operation is required, requested name:%s",operation), null);
+        }
+        
+        ZLdapContext zlc = null;
+        Element response = zsc.createElement(AdminConstants.HAB_ORG_UNIT_RESPONSE);
+        try {
+            zlc = LdapClient.getContext(LdapServerType.MASTER, LdapUsage.CREATE_OU);
+
+            ZAttributes ouAttrs = null;
+            switch (operation) {
+            case "create":
+                
+                String[] objClass = ArrayUtils.toStringArray(LdapObjectClass.getOrganizationUnitObjectClasses().toArray());
```

Comment:
create/delete/rename ou implementation is already there in ldap provisioning.so, we should use those methods here.